runtime-common/current-run.ts

```diff
@@ -0,0 +1,769 @@
+
+  private constructor({
+    realm,
+    reader,
+    instances,
+    modules,
+    definitions,
+    ignoreMap,
+    exportedCardRefs,
+    incremental,
+  }: {
+    realm: Realm;
+    reader: Reader | undefined; // the "empty" case doesn't need a reader
+    instances: URLMap<SearchEntry>;
+    modules: URLMap<ModuleSyntax>;
+    definitions: Map<string, CardDefinition>;
+    ignoreMap: URLMap<Ignore>;
+    exportedCardRefs: Map<string, Map<string, ExportedCardRef>>;
+    incremental: { url: URL; operation: "update" | "delete" } | undefined;
+  }) {
+    this.#realmPaths = new RealmPaths(realm.url);
+    this.#reader = reader;
+    this.realm = realm;
+    this.#instances = instances;
+    this.#modules = modules;
+    this.#definitions = definitions;
+    this.incremental = incremental;
+    this.#exportedCardRefs = exportedCardRefs;
+    this.#ignoreMap = ignoreMap;
+  }
+
+  static empty(realm: Realm) {
+    return new this({
+      realm,
+      reader: undefined,
+      instances: new URLMap(),
+      modules: new URLMap(),
+      definitions: new Map(),
+      incremental: undefined,
+      exportedCardRefs: new Map(),
+      ignoreMap: new URLMap(),
+    });
+  }
+
+  static async fromScratch(realm: Realm, reader: Reader) {
+    let current = new this({
+      realm,
+      reader,
+      instances: new URLMap(),
+      modules: new URLMap(),
+      // TODO need to investigate the syntax analysis of the base card--if that
+      // is not possible to do, then we should seed it here
+      definitions: new Map([
+        // seed the definitions with the base card
+        [
+          internalKeyFor(
+            {
+              type: "exportedCard",
+              module: `${baseRealm.url}card-api`,
+              name: "Card",
+            },
+            undefined
+          ),
+          {
+            id: {
+              type: "exportedCard",
+              module: `${baseRealm.url}card-api`,
+              name: "Card",
+            },
+            key: internalKeyFor(
+              {
+                type: "exportedCard",
+                module: `${baseRealm.url}card-api`,
+                name: "Card",
+              },
+              undefined
+            ),
+            super: undefined,
+            fields: new Map(),
+          },
+        ],
+      ]),
+      incremental: undefined,
+      exportedCardRefs: new Map(),
+      ignoreMap: new URLMap(),
+    });
+    await current.run();
+    return current;
+  }
+
+  static async incremental(
+    url: URL,
+    operation: "update" | "delete",
+    prev: CurrentRun
+  ) {
+    let instances = new URLMap(prev.instances);
+    let modules = new URLMap(prev.modules);
+    let exportedCardRefs = new Map(prev.exportedCardRefs);
+    let ignoreMap = new URLMap(prev.ignoreMap);
+    let definitions = new Map(prev.definitions);
+    // TODO refactor this away via invalidation mechanism...
+    removeDefinitions(url, definitions);
```

Comment:
this is suboptimal--this should be refactored away as part of invalidation working correctly